# translationError_ageing
## Prerequisites:
* Python 2 
* MaxQuant v2.7.0.0
* Cutadapt v.1.8
* bowtie 1.3.1

## Analysis pipeline
### 1 Detecting misincorporation rate of amino acids
1.1 Raw mass spectrometry (MS) data processing
* Input:
  * Raw MS data downloaded from the Proteomics Identification Database, with the accession numbers of all used samples listed in TableS1
  
```{shell}
# Taking yeast MS data as an example, use MaxQuant to generate mqpar.xml
dotnet /path/MaxQuantCmd.exe mqpar.xml
```
* Output
  * dependentPeptides.txt in txt folder generated by MaxQuant,takes about 20 hours on a "normal" desktop computer.
    
1.2 Identification of misincorporation events
* Input 
  * dependentPeptides.txt
```{shell}
python2 detect_quant.py
```
* Output
  * detec_quant.csv, takes about 20 mins on a "normal" desktop computer.
    
1.3 Calculation of misincorporation rates in young and old individuals

```{r}
aa_subRate_calculation.R

```

### 2.Detecting translational readthrough
2.1 Sequence alignment
* Input
  * Raw Ribo-Seq data were downloaded from GEO, with the accession numbers listed in TableS2.
```{shell}
bash sequence_alignment.sh

```
* Output:
  * bam files in align folder generated by sequence_alignment.sh, takes about 5 hours on a "normal" desktop computer.
  
2.2 readthrough efficiency calculation

* Input
  * bam files in align folder 
```{shell}
readthrough_efficiency_calculation.R
```
* Output:
  * readcount.RData, takes about 10 mins on a "normal" desktop computer.
  
