# translationError_ageing
##Prerequisites: 

*Python 2 
*MaxQuant v2.7.0.0
*Cutadapt v.1.8
*bowtie 1.3.1

##Analysis pipeline
###1 Detecting misincorporation rate of amino acids

1.1 Raw mass spectrometry (MS) data processing
*Input:
  *Raw MS data downloaded from the Proteomics Identification Database, with the accession numbers of all used samples listed in Table_S1
  
```{shell}
#Taking yeast MS data as an example, use MaxQuant to generate mqpar.xml
dotnet /path/MaxQuantCmd.exe mqpar.xml
```
*Output
  *dependentPeptides.txt in txt folder generated by MaxQuant
1.2 Identification of misincorporation events
*Input 
  *dependentPeptides.txt
```{shell}
#Taking human MS data as an example, use MaxQuant to generate mqpar_liunx.xml
python2 detect_quant.py
```
*Output
  *detec_quant.csv
1.3 Calculation of misincorporation rates in young and old individuals

```{r}
aa_subRate_calculation.R

```


###2.Detecting translational readthrough
2.1 Sequence alignment
*Input
  *Raw Ribo-Seq data were downloaded from GEO, with the accession numbers listed in Table S2.
```{shell}
bash sequence_alignment.sh

```
*Output:
  *bam files in align folder generated by sequence_alignment.sh
  
2.2 readthrough efficiency calculation

*Input
  *Rbam files in align folder 
```{shell}
readthrough_efficiency_calculation.R
```
*Output:
  *readcount.RData
  
